### DISCUSSION TITLE
如何使用Go实现一个简单的负载均衡器

### DISCUSSION THREAD

**Discussion by beijixi** [RunwangGuo/short-url-4go#3] [Q&A] [answered]
如何使用Go实现一个简单的负载均衡器

**Comment by RunwangGuo** [comment, marked as answer]
以下是一个简单的基于轮询的负载均衡器实现:
package main

import (
 "fmt"
 "log"
 "net/http"
 "sync"
)

type Backend struct {
 URL string
 Weight int
}

type LoadBalancer struct {
 backends []Backend
 mu sync.Mutex
 currentIndex int
}

func NewLoadBalancer(backends ...Backend) *LoadBalancer {
 lb := &LoadBalancer{backends: backends}
 lb.currentIndex = 0
 return lb
}

func (lb *LoadBalancer) RoundRobin() string {
 lb.mu.Lock()
 defer lb.mu.Unlock()

 index := lb.currentIndex % len(lb.backends)
 lb.currentIndex++

 return lb.backends[index].URL
}

func (lb *LoadBalancer) AddBackend(url string, weight int) {
 lb.mu.Lock()
 defer lb.mu.Unlock()

 for i, backend := range lb.backends {
  if backend.URL == url {
   lb.backends[i].Weight += weight
   return
  }
 }

 lb.backends = append(lb.backends, Backend{URL: url, Weight: weight})
}

func (lb *LoadBalancer) RemoveBackend(url string) {
 lb.mu.Lock()
 defer lb.mu.Unlock()

 for i, backend := range lb.backends {
  if backend.URL == url {
   lb.backends = append(lb.backends[:i], lb.backends[i+1:]...)
   return
  }
 }
}

func main() {
 lb := NewLoadBalancer(
  Backend{"http://backend1.example.com", 1},
  Backend{"http://backend2.example.com", 1},
  Backend{"http://backend3.example.com", 1},
 )

 http.HandleFunc("/proxy", func(w http.ResponseWriter, r *http.Request) {
  target := lb.RoundRobin()
  fmt.Fprintf(w, "Proxying request to: %s\n", target)
 })

 log.Fatal(http.ListenAndServe(":8080", nil))
}
